code/modules/projectiles/guns/ballistic.dm

```diff
@@ -301,13 +322,30 @@
 	if (can_be_sawn_off)
 		if (sawoff(user, A))
 			return
+	
+	if(can_misfire && istype(A, /obj/item/stack/sheet/cloth))
+		if(guncleaning(user, A))
+			return
+	
 	return FALSE
 
 /obj/item/gun/ballistic/process_fire(atom/target, mob/living/user, message = TRUE, params = null, zone_override = "", bonus_spread = 0)
+
+	if(magazine && chambered.BB && can_misfire && misfire_probability > 0)
+		if(prob(misfire_probability))
+			if(blow_up(user))
+				to_chat(user, "<span class='userdanger'>[src] misfires!</span>")
+
 	if (sawn_off)
 		bonus_spread += SAWN_OFF_ACC_PENALTY
 	. = ..()
 
+/obj/item/gun/ballistic/shoot_live_shot(mob/living/user, pointblank = 0, atom/pbtarget = null, message = 1)
+	if(can_misfire)
+		misfire_probability += misfire_percentage_increment
```

Comment:
I forget, does shoot_live_shot happen before or after process_fire?

Reply:
If I remember correctly since this was my concern at the time, it's after. So process_fire calls shoot_live_shot if it would fire a shot (and so wouldn't call it if it didn't have a live shot, therefore doesn't foul if it would be empty for example).

I tested this with having it foul with just process_fire, and the result was a gun increasing misfire chance when empty, so best not to do that.

Reply:
I double checked this and I am fairly confident that process_fire isn't going to cause det revolver to detonate before the first shot given the procedures. I also couldn't get it to misfire on the first shot after a cleaning, so it appears it is working as intended.